lib/modules.nix

```diff
@@ -424,12 +435,66 @@ rec {
         else if tp.name == "list of option sets" then types.listOf (types.submodule options)
         else if tp.name == "null or option set" then types.nullOr (types.submodule options)
         else tp;
+
     in
       if opt.type.getSubModules or null == null
-      then opt // { type = f (opt.type or types.unspecified); }
+      then opt // { type = f (opt.type or (inferType typeInference loc opt)); }
```

Comment:
@nbp My thought was that `_module.typeInference` could be used to disable the type inference, and so a performance hit, for an ordinary evaluation. Do you think that's not necessary?